\documentclass[fleqn,titlepage,11pt]{article}

\usepackage{latexsym,delcher}

\PortraitPage
\def\baselinestretch{1.0}
\def\thefootnote{\fnsymbol{footnote}}
\def\thepage{{\footnotesize\arabic{page}}}
\def\today{17~May~2011}

\def\Desc#1{\,\mbox{\emph{#1}}\,}
\def\Glimmer{\textsc{Glimmer}}
\def\Gtwo{\textsc{Glimmer2}}
\def\Gthree{\textsc{Glimmer3}}
\def\Gmg{\textsc{Glimmer-MG}}
\def\Phymm{\textsc{Phymm}}
\def\Scimm{\textsc{Scimm}}
\def\PhyScimm{\textsc{PhyScimm}}
\def\PgBICM{\texttt{build-icm}}
\def\Pg#1{\texttt{#1}}


\begin{document}

\RaggedRight
\sloppy

\title{\Gmg{} Release Notes \\ Version~0.1}
\author{David R. Kelley, Art L. Delcher}
\titlepagenote{Copyright \copyright\ 2011 University of Maryland Center for Bioinformatics \& Computational Biology}

\maketitle

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Introduction}

This document describes Version~0.1 of the \Gmg{} gene-finding
software for metagenomic sequences.  Users discovering problems or
errors are encouraged to report them to
\,\verb`dakelley@umiacs.umd.edu`\,.

\Glimmer{} is a collection of programs for identifying genes in
microbial DNA sequences.  The system works by creating a
variable-length Markov model from a training set of genes and then
using that model to attempt to identify all genes in a given DNA
sequence.  The three versions of \Glimmer{} are described
in~\cite{glimmer1},~\cite{glimmer2}, and~\cite{glimmer3}.

\Gmg{} is released as OSI Certified Open Source Software under the
Artistic License.  The license is contained in the file, \Pg{LICENSE},
in the distribution.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Installation}

\Gmg{} software was written for the Linux software environment.  The
following instructions assume a Linux system.  They also work under
Mac OSX.

To install \Gmg{}, first, download and install BioPython
(http://biopython.org/wiki/Biopython). Then, download the compressed
tarfile \,\verb`glimmer-mg-1.0.1.tar.gz`\, from the website.  Then
uncompress the file by typing \BSV
\begin{verbatim}
  tar xzf glimmer-mg.1.0.1.tar.gz
\end{verbatim}
\ESV

A directory named \,\verb`glimmer-mg`\, should result.
In that directory is a script called \,\verb`install_glimmer.py`\,
which will download and install the prerequisite software and
compile the \Gmg{} code. More specifically, it will
\begin{enumerate}\RaggedRight

\item Download, compile, and install \,\verb`Phymm`\,

  \,\verb`Phymm`\, is used to compute the phylogenetic classifications
  of sequences used by \Gmg{} to parameterize gene prediction models.
  \,\verb`Phymm`\, will install a database of interpolated Markov
  models (IMMs) to be used for classification that requires $\sim50$
  Gb of disk space and takes about $\sim24$ hours to build. If you
  already have \,\verb`Phymm`\, installed, you need only set the
  variable \Desc{prior\_phymm\_dir} in \,\verb`install_glimmer.py`\, to
  its installation directory and set \Desc{install\_phymm} to False.

\item Download, compile, and install \,\verb`Scimm`\,

  \,\verb`Scimm`\, is used to cluster the sequences to enable an
  unsupervised retraining step followed by a second iteration of
  gene predictions.

\item Download, compile, and install \,\verb`ELPH`\,

  \,\verb`ELPH`\, is used to compute a motif position weight matrix
  for the ribosomal binding sites in the promoters of the training
  gene sets.

\item Compile and install \Gmg{}

  This will compile the main gene prediction code.

\item Prepare \Gmg{} training data

  There are three steps here. First, the script
  \,\verb`train_all.py`\, will train the models for all features on
  every reference genome in the \,\verb`Phymm`\, database. Second, the
  script \,\verb`informative_genomes.py`\, will decide which reference
  genomes will be useful for gene prediction and which (like small
  plasmids) will not be useful. Finally, the script
  \,\verb`double_icms.py`\, will determine which ICMs trained on two
  reference genomes are worth building and build them. These steps
  require some time.

\end{enumerate}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Running \Gmg{}}
\Gmg{} can be run in a few different modes, depending on the
characteristics of the sequences provided, e.g., whether they
represent accurate contigs or reads from the Illumina or 454
technologies. Then within each mode, \Gmg{} can be run with varying
amounts of classification/clustering preprocessing.

\subsection{Preprocessing options}
The \Gmg{} pipeline can be run a number of different ways, including a
Python script to run the full pipeline and options to classify/cluster
the reads separately and predict genes using those results.

% Running on a single genome
\subsubsection{Running on a single genome}
The Python script \,\Pg{g3-iterated.py}\, runs the unsupervised single
genome pipeline.
\BSV
\begin{verbatim}
g3-iterated.py genome.fasta out
\end{verbatim}
\ESV
This script will extract initial ORFs for training using the \,\Pg{long-orfs}\,
program and then run multiple iterations of Glimmer and retraining.

% Running everything
\subsubsection{Running from scratch}
The Python script \,\Pg{glimmer-mg.py}\, runs the full \Gmg{}
pipeline.
\BSV
\begin{verbatim}
glimmer-mg.py seqs.fasta
\end{verbatim}
\ESV
This script will classify the sequences with \Phymm{}~\cite{phymm},
make initial predictions for all sequences, cluster the sequences with
\Scimm{}~\cite{scimm}, and make final predictions within each cluster.

% Running everything, no re-predictions
\subsubsection{Running from scratch with no retraining}
By passing the option \,\Pg{--iter 0}\, to \,\Pg{glimmer-mg.py}\,, the
clustering and re-training steps will be skipped. That is, the
sequences will be classified with \Phymm{} and predictions made.
\BSV
\begin{verbatim}
glimmer-mg.py --iter 0 seqs.fasta
\end{verbatim}
\ESV

% Classifying first, then running
\subsubsection{Classification separate}
Some users may prefer to run the computationally intensive
classification of sequences separately (e.g. on a computer cluster).
\,\Pg{glimmer-mg.py}\, can be made to expect this by using the
\,\Pg{--raw}\, option, which specifies that the raw \Phymm{} output file
already exists in the current working directory.
\BSV
\begin{verbatim}
glimmer-mg.py --raw seqs.fasta
\end{verbatim}
\ESV

% Classifying and clustering first, then running
\subsubsection{Clustering separate}
In a similar vein, clustering can be performed separately before
making gene predictions by specifying the \,\Pg{--clust}\, option,
which specifies that \Scimm{} or \PhyScimm{} output is in the current
working directory.
\BSV
\begin{verbatim}
glimmer-mg.py --clust seqs.fasta
\end{verbatim}
\ESV

\subsection{Sequence modes}
Depending on the error characteristics of the input sequences, \Gmg{}
has three modes that best handle certain types of data. In each case,
the options described can be passed to the C++ binary
\,\Pg{glimmer-mg}\, or the Python script to manage the entire pipeline
\,\Pg{glimmer-mg.py}\,.

\subsubsection{Accurate contigs}
If you believe that the sequences on which you would like to find
genes are very accurate, than run \Gmg{} in the default mode with no
additional options.

\subsubsection{454 indels}
If the input sequences are reads or contigs from the 454 or Ion
Torrent technologies and indel errors are expected to exist, \Gmg{}
can be made to predict such errors and account for them in its gene
predictions by using \emph{indel} mode.
\BSV
\begin{verbatim}
glimmer-mg.py --indels 454.fasta
\end{verbatim}
\ESV

\subsubsection{Illumina substitution errors}
If the input sequences are reads or contigs from the Illumina
technology and substitution errors are expected to exist, \Gmg{}
can be made to predict such errors and account for them in its gene
predictions by using \emph{substitution} mode.
\BSV
\begin{verbatim}
glimmer-mg.py --sub illumina.fasta
\end{verbatim}
\ESV

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Sample Run Directory}
The directory \,\Pg{sample\_run}\, contains a sample run of \Gmg{}.

\subsection{Single Genome}
The subdirectory \,\Pg{glimmer3}\, contains the genome sequence of
a Helicobacter Pylori strain. The files in the directory \,\Pg{results}\, are the
result of running the script
\BSV
\begin{verbatim}
  g3-iterated.py NC_000915.fna NC_000915
\end{verbatim}
\ESV

\subsection{Metagenomics}
The subdirectory \,\Pg{glimmer-mg}\, contains a simulated metagenome
in the fasta file \,\Pg{seqs.fa}\, with the sequence origins in
\,\Pg{map.txt}\,. The files in the directory \,\Pg{results}\, are the
result of running the script
\BSV
\begin{verbatim}
  glimmer-mg.py seqs.fa
\end{verbatim}
\ESV

Don't be alarmed if your results are slightly different. One
explanation is that a slightly different set of GenBank reference
genomes were available to \Phymm. Another explanation is that \Scimm{}
has a stochastic component and may have produced a different set of
clusters used for retraining.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Notes on \Gmg{} programs}
\subsection{\Pg{glimmer-mg}}

This is the main program that makes gene predictions.

\subsubsection{\Pg{glimmer-mg} Parameters \& Options}
The invocation for \,\Pg{glimmer-mg}\, is:
\bq
  \Pg{glimmer-mg}\, [\Desc{options}] \Desc{sequence} \Desc{tag}
\eq
where \Desc{sequence} is the name of the file containing the DNA
sequence(s) to be analyzed and \Desc{tag} is a prefix used to name the
output files.

\Desc{options} can be the following:
\bl{}\RaggedRight

\exdent
  \verb`-b` \Desc{filename} \enskip or \enskip \verb`--rbs_pwm` \Desc{filename}

  Read a position weight matrix (PWM) from \Desc{filename} to identify
  the ribosome binding site to help choose start sites.  The format of
  this file is indicated by the following example:
\BSV
\begin{verbatim}
6
a     212     309      43      36     452     138
c      55      58       0      19      48      26
g     247     141     501     523       5     365
t      64      70      34       0      73      49
\end{verbatim}
\ESV
  The first line is the number of positions in the pattern, \ie,
  the number of columns in the matrix (not counting
  the first column of labels).  The column values are the relative
  frequencies of nucleotides at each position.

\exdent
  \verb`-c` \Desc{filename} \enskip or \enskip \verb`--class` \Desc{filename}

  Read sequence classifications from \Desc{filename}. Each line of
  \Desc{filename} should specify a sequence fasta header followed by a
  series of GenBank reference genome classifications. The
  classifications should name the reference genome's Phymm directory
  (corresponding to the GenBank strain name), followed by the
  character '$|$', followed by the genomic sequence's unique
  identifier. It is our expectation that users will not create these
  files, but they generate them with \,\Pg{glimmer-mg.py}\,. For
  example:

\BSV
\footnotesize{
\begin{verbatim}
read1     Mycobacterium_leprae_Br4923|NC_011896 Corynebacterium_glutamicum_ATCC_13032|NC_003450
read2     Rhodopseudomonas_palustris_BisA53|NC_008435 Rhodopseudomonas_palustris_HaA2|NC_007778
...
\end{verbatim}
}
\ESV

\exdent
  \verb`-f` \Desc{filename} \enskip or \enskip \verb`--features` \Desc{filename}

  \Desc{filename} specifies counts for features such as gene length,
  start codon usage, adjacent gene orientations, and adjacent gene
  distances for coding and noncoding ORFs. \Gmg{} will convert these
  counts into probability models for use in log-likelihood
  computations for the features. It is our expecation that users will
  not create these files, but generate them with
  \,\Pg{train\_features.py}\, (via \,\Pg{glimmer-mg.py}\,). For an
  example of the input format, see the \,\Pg{sample\_run}\, directory.

\exdent
  \verb`-g` \Desc{n} \enskip or \enskip \verb`--gene_len` \Desc{n}

  Set the minimum gene length to \Desc{n} nucleotides.  This does not include
  the bases in the stop codon.

\exdent
  \verb`-h` \enskip or \enskip \verb`--help`

  Print the usage message.

\exdent
  \verb`-i` \enskip or \enskip \verb`--indel`

  Predict genes in ``indel-mode'' where gene predictions may shift the
  coding frame, implicitly predicting an insertion or deletion in the
  sequence. If quality values are provided with the \verb`-q` option,
  positions at which frame shifts should be considered will be
  identifed by their low quality. If no quality values are provided,
  frame shifts will be considered at long homopolymer runs, assuming
  that the sequences were generated by the 454 or Ion torrent
  technologies.

\exdent
  \verb`-m` \Desc{filename} \enskip or \enskip \verb`--icm` \Desc{filename}

  \Desc{filename} contains an ICM trained on coding sequence to be used
  for ORF scoring rather than ICM's obtained via classification results.

\exdent
  \verb`-o` \Desc{n} \enskip or \enskip \verb`--max_olap` \Desc{n}

  Set the maximum overlap length to \Desc{n}.  Overlaps of this
  many or fewer bases are allowed between genes.  The new
  dynamic programming algorithm should \underline{\emph{never}}
  output genes that overlap by more than this many bases.

\exdent
  \verb`-q` \Desc{filename} \enskip or \enskip \verb`--quality` \Desc{filename}

  \Desc{filename} contains quality values in fasta format matching up
  with the sequences fasta file. The quality values are used in
  ``indel-mode'' and ``substitution-mode'' to identify low quality
  positions in the sequences where errors are most likely.

\exdent
  \verb`-r` \enskip or \enskip \verb`--circular`

  Assume a circular rather than linear genome, \ie, there may
  be genes that ``wraparound'' between the beginning and end
  of the sequence.

\exdent
  \verb`-s` \enskip or \enskip \verb`--sub`

  Predict genes in ``substitution-mode'' where gene predictions may
  pass through stop codons, implicitly predicting a sequencing error
  that mutated a standard codon to a stop codon. If quality values are
  provided with the \verb`-q` option, they will be used to compute a
  log-likelihood penalty for passing through a given stop
  codon. Otherwise, we assume all quality values are Phred 30.

\exdent
  \verb`-u` \Desc{n} \enskip or \enskip \verb`--fudge` \Desc{n}

  \Desc{n} specifices a ``fudge factor'' to be added to the
  log-likeihood score for every ORF. The ``fudge factor'' acts as a
  means to tune the sensitivity versus specificity of the predictions.
  
\exdent
  \verb`-z` \Desc{n} \enskip or \enskip \verb`--trans_table` \Desc{n}

  Use Genbank translation table number \Desc{n} to specify stop codons.

\exdent
  \verb`-Z` \Desc{codon-list} \enskip or \enskip \verb`--stop_codons` \Desc{codon-list}

  Specify stop codons as a comma-separated list.
  Sample format:  \verb`-Z tag,tga,taa`.
  The default stop codons are \Pg{tag}, \Pg{tga} and \Pg{taa}.
\el


\subsubsection{\Pg{glimmer-mg} Output Formats}

\smallskip
\noindent\textbf{\Pg{.predict} File}
\smallskip

This file has the final gene predictions.  It's format is the fasta-header
line of the sequence followed by one line per gene.  Here is a sample of the
beginning of such a file:
\BSV
\begin{verbatim}
>SRR029690.117009 E4LJNJL02B7ZT9 length=531
orf00045      386       -1  -2    52.14 I:272 D: S:
orf00057      532      450  -3     3.86 I: D:507 S:
\end{verbatim}
\ESV
The columns are:
\bl{\settowidth{\labelwidth}{Column 1}\leftmargin=\labelwidth \addtolength{\leftmargin}{1em}\labelsep=1em}\RaggedRight
\item[Column 1]
  The identifier of the predicted gene.

\item[Column 2]
  The start position of the gene.

\item[Column 3]
  The end position of the gene.  This is the last base of the stop codon, \ie,
  it includes the stop codon.

\item[Column 4]
  The reading frame.

\item[Column 5]
  The log-likelihood ratio score of the gene.

\item[Column 6]
  The positions of insertion, deletion, and substition predictions within the ORF.
\el

\subsection{\Pg{extract\_aa.py}}
\,\Pg{extract\_aa.py}\, can be used to extract a multi-fasta file of
amino acid sequences from the gene predictions made by
\,\Pg{Glimmer}\,.

\subsubsection{\Pg{extract\_aa.py} Parameters \& Options}

The invocation for \,\Pg{extract\_aa.py}\, is:
\bq
  \Pg{extract\_aa.py}\, [\Desc{options}]
\eq

\Desc{options} can be the following:
\bl{}\RaggedRight

\exdent
  \verb`-s` \Desc{filename}

  \Desc{filename} specifies the fasta file of sequences on which
  gene predictions were made.

\exdent
  \verb`-p` \Desc{filename}

  \Desc{filename} specifies the Glimmer output .predict file
  containing gene predictions.

\exdent
  \verb`-o` \Desc{filename}

  \Desc{filename} specifies the output file in which to print
  amino acid sequences in fasta format.

\el

\subsection{\Pg{informative\_genomes.py}}

Make a list of the GenBank reference genomes which have sufficient
non-hypothetical gene annotations to be useful for training. Only
Phymm classifications to these genomes will be used for training.

\subsection{\Pg{train\_features.py}}

Generate data files to be directly used for gene prediction training.

\subsubsection{\Pg{train\_features.py} Parameters \& Options}

The invocation for \,\Pg{train\_features.py}\, is:
\bq
  \Pg{train\_features.py}\, [\Desc{options}]
\eq

\Desc{options} can be the following:
\bl{}\RaggedRight

\exdent
  \verb`-f`

  Print a ``feature'' file as expected by the \verb`-f` option of the
  \Pg{glimmer-mg}

\exdent
  \verb`--gbk` \Desc{filename}

  \Desc{filename} specifies the GenBank RefSeq annotation file from
  which to generate data files. Either this option or the combination
  of \verb`--predict` and \verb`--seq` must be specified.

\exdent
  \verb`--icm`

  Only produce the gene composition ICM model.

\exdent
  \verb`--indels`

  Note that the gene predictions may contain indels, which must be
  carefully considered when processing the predictions.

\exdent
  \verb`-l` \Desc{n}

  \Desc{n} specifies the minimum length of an ORF to be considered
  as a gene.

\exdent
  \verb`--predict` \Desc{filename}

  \Desc{filename} specifies \,\Pg{Glimmer}\, gene prediction output to
  be used for training. \verb`--seq` must also be specified. Either
  this option or \verb`--gbk` must be specified.

\exdent
  \verb`-o` \Desc{n} \enskip or \enskip \verb`--max_overlap` \Desc{n}

\exdent
  \verb`--rbs`

  Only produce the ribosomal binding site motif model.

\exdent
  \verb`--seq` \Desc{filename}

\exdent
  \verb`-z`

\el

\subsubsection{\Pg{train\_features.py} Output Formats}

\smallskip
\noindent\textbf{\Pg{.gicm} File}
\smallskip

Interpolated context model (ICM) trained on coding sequence from the
genome given.

\smallskip
\noindent\textbf{\Pg{.gc.txt} File}
\smallskip

GC\% of the genome given.

\smallskip
\noindent\textbf{\Pg{.lengths.<genes/non>.txt} File}
\smallskip

Counts of gene or noncoding ORF lengths from the genome given.

\smallskip
\noindent\textbf{\Pg{.starts.<genes/non>.txt} File}
\smallskip

Counts of start codon usage for the genes and noncoding ORFs
from the genome given.

\smallskip
\noindent\textbf{\Pg{.motif} File}
\smallskip

Ribosomal binding site (RBS) position weight matrix motif model
as generated by ELPH from the promoters of genes in the genome
given.

\smallskip
\noindent\textbf{\Pg{.adj\_orients.<genes/non>.txt} File}
\smallskip

Counts of gene or noncoding adjacent gene orientations. $1$ specifies
an ORF in the forward direction, and $-1$ specifies an ORF in the
revere direction. Thus, $1,-1$ specifies the number of times a forward
gene is followed by a reverse gene.

\smallskip
\noindent\textbf{\Pg{.adj\_dist.<1/-1>.<1/-1>.<genes/non>.txt} File}
\smallskip

Counts of distances between adjacent genes or noncoding ORFs. $1$
specifies an ORF in the forward direction, and $-1$ specifies an ORF
in the revere direction. Thus, $1.-1$ specifies the distance counts
when a forward gene is followed by a reverse gene. In the noncoding
case, the distance counts refer to each pair of a noncoding ORF and
the genes adjacent to it.

\subsection{\Pg{train\_all.py}}

Run \Pg{train\_features.py} on all GenBank reference genomes in
Phymm's database.

\subsubsection{\Pg{train\_all.py} Parameters \& Options}

The invocation for \,\Pg{train\_all.py}\, is:
\bq
  \Pg{train\_all.py}\, [\Desc{options}]
\eq

\Desc{options} can be the following:
\bl{}\RaggedRight

\exdent
  \verb`-p` \Desc{n}

  Spread the processes launched over \Desc{n} processes.

\exdent
  \verb`-u`

  Only train on GenBank reference genomes for which no data files
  have yet been produced.

\el

\subsection{\Pg{double\_icms.py}}

Generate ICMs trained on pairs of GenBank reference genome
annotations. Only pairs of reference genomes with siimilar composition
are generated. The similarity function is defined in the \Gmg{}
manuscript.



%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Notes on \Glimmer{} programs}

\subsection{\Pg{build-icm} Program}

This program constructs an interpolated context model (ICM)
from an input set of sequences.

\subsubsection{\Pg{build-icm} Parameters \& Options}
The format for invoking \,\Pg{build-icm}\, is:
\bq
  \Pg{build-icm}\, [\Desc{options}] \Desc{output-file} \,\Pg{<}\,\Desc{input-file}
\eq
Sequences are reads from standard input, the ICM is
built and written to \Desc{output-file}.  If \Desc{output-file}
is ``-'', then the output will be sent to standard output.
Since input comes from standard input, one also can ``pipe'' the input
into this program, \eg,
\BSV
\begin{verbatim}
  cat abc.in | build-icm xyz.icm
\end{verbatim}
\ESV
or even type in the input directly.

Possible \Desc{options} are:
\bl{}\RaggedRight
\exdent
  \verb`-d` \Desc{num} \enskip or \enskip \verb`--depth` \Desc{num}

  Set the depth of the ICM to \Desc{num}.  The depth is the
  maximum number of positions in the context window that
  will be used to determine the probability of the predicted
  position.  The default value is 7.

\exdent
  \verb`-F` \enskip or \enskip \verb`--no_stops`

  Do not use any input strings with in-frame stop codons.
  Stop codons are determined by either the \Pg{-z} or \Pg{-Z}
  option.

\exdent
  \verb`-h` \enskip or \enskip \verb`--help`

  Print the usage message.

\exdent
  \verb`-p` \Desc{num} \enskip or \enskip \verb`--period` \Desc{num}

  Set the period of the ICM to \Desc{num}.  The period is the
  number of different submodels for different positions in the
  text in a cyclic pattern.  \Eg, if the period is 3, the first
  submodel will determine positions $1, 4, 7, \dots$; the second
  submodel will determine positions $2, 5, 8, \dots$; and the third
  submodel will determine positions $3, 6, 9, \dots$.  For a
  non-periodic model, use a value of 1.  The default value
  is 3.

\exdent
  \verb`-r` \enskip or \enskip \verb`--reverse`

  Use the reverse of the input strings to build the ICM.  Note that
  this is merely the reverse and \emph{\underline{NOT}} the
  reverse-complement.  In other words, the model is built in
  the backwards direction.

\exdent
  \verb`-t` \enskip or \enskip \verb`--text`

  Output the model in a text format.  This is for
  informational/debugging purposes only---the \Pg{glimmer3}
  program cannot read models in this form.

  The format of the output is a header line containing the
  parameters of the model, followed by individual
  probability lines.  The entries on each probability line
  are:
  \bq
    \begin{tabular}{cl}
      Column & \quad Description \\
      1 & ID number \\
      2 & Context pattern \\
      3 & Mutual information \\
      4 & Probability of A \\
      5 & Probability of C \\
      6 & Probability of G \\
      7 & Probability of T
    \end{tabular}
  \eq
  The context pattern is divided into codons by the vertical lines (this
  option assumes the default 3-periodic model).
  The ``?'' represents the position being predicted.  Letters represent
  specific values in their respective positions in the context window.
  The asterisk indicates the position that has maximum mutual information
  with the predicted position.

\exdent
  \verb`-v` \Desc{num} \enskip or \enskip \verb`--verbose` \Desc{num}

  Set the verbose level to \Desc{num}.  This controls extra debugging
  output---the higher the value the more output.

\exdent
  \verb`-w` \Desc{num} \enskip or \enskip \verb`--width` \Desc{num}

  Set the width of the ICM to \Desc{num}.  The width includes
  the predicted position.  The default value is 12.

\exdent
  \verb`-z` \Desc{n} \enskip or \enskip \verb`--trans_table` \Desc{n}

  Use Genbank translation table number \Desc{n} to specify stop codons.

\exdent
  \verb`-Z` \Desc{codon-list} \enskip or \enskip \verb`--stop_codons` \Desc{codon-list}

  Specify stop codons as a comma-separated list.
  Sample format:  \,\verb`-Z tag,tga,taa`\,.
  The default stop codons are \Pg{tag}, \Pg{tga} and \Pg{taa}.
\el

\subsection{\Pg{long-orfs} Program}

This program identifies long, non-overlapping open reading frames (orfs)
in a DNA sequence file.  These orfs are very likely to contain genes,
and can be used as a set of training sequences for the \Pg{build-icm}
program.  More specifically, among all orfs longer than a minimum length
$\ell$, those that do not overlap any others are output.  The start
codon used for each orf is the first possible one.  The program, by
default, automatically determines the value $\ell$ that maximizes the
number of orfs that are output.  With the \Pg{-t} option, the initial
set of candidate orfs also can be filtered using entropy distance, which
generally produces a larger, more accurate training set, particularly
for high-GC-content genomes.  Entropy distance is described in~\cite{med1}.

\subsubsection{\Pg{long-orfs} Parameters \& Options}
The format for invoking \,\Pg{long-orfs}\, is:
\bq
  \Pg{long-orfs}\, [\Desc{options}] \Desc{sequence} \Desc{output}
\eq
where \Desc{sequence} is the name of the file containing the DNA sequence
to be analyzed and \Desc{output} is the name of the output file of
coordinates.  \Desc{sequence} may contain only one sequence.
If \Desc{output} is ``\Pg{-}'', then the output is directed to
standard output.

Possible \Desc{options} are:
\bl{}\RaggedRight
\exdent
  \verb`-A` \Desc{codon-list} \enskip or \enskip \verb`--start_codons` \Desc{codon-list}

  Specify allowable start codons as a comma-separated list.
  Sample format:  \,\verb`-A atg,gtg`\,.
  The default start codons are \Pg{atg}, \Pg{gtg} and \Pg{ttg}.

\exdent
  \verb`-E` \Desc{filename} \enskip or \enskip \verb`--entropy` \Desc{filename}

  Read entropy profiles from \Desc{filename}.  The format is one header
  line, then 20 lines of 3 columns each, which is the format produced
  by the program \Pg{entropy-profile} with the \Pg{-b} option.
  The columns are amino acid,
  positive entropy, and negative entropy, respectively.  Rows must be in
  alphabetical order by amino acid code letter.

  The entropy profiles are used only if the \Pg{-t} option is specified.

\exdent
  \verb`-f` \enskip or \enskip \verb`--fixed`

  Do \underline{\emph{NOT}} automatically calculate the minimum gene
  length that maximizes the number or length of output regions, but
  instead use either the value specified by the \Pg{-g} option or
  else the default, which is 90.

\exdent
  \verb`-g` \Desc{n} \enskip or \enskip \verb`--min_len` \Desc{n}

  Set the minimum gene length to \Desc{n} nucleotides.  This does not include
  the bases in the stop codon.

\exdent
  \verb`-h` \enskip or \enskip \verb`--help`

  Print the usage message.

\exdent
  \verb`-i` \Desc{filename} \enskip or \enskip \verb`--ignore` \Desc{filename}

  \Desc{filename} specifies regions of bases that are off 
  limits, so that no bases within that area will be examined.
  The format for entries in this file is described above for
  the same option in the \Pg{glimmer3} program.

\exdent
  \verb`-l` \enskip or \enskip \verb`--linear`

  Assume a linear rather than circular genome, \ie, there will
  be no ``wraparound'' genes with part at the beginning of the sequence
  and the rest at the end of the sequence.

\exdent
  \verb`-L` \enskip or \enskip \verb`--length_opt`

  Find and use as the minimum gene length the value that maximizes the
  total \underline{\emph{length}} of non-overlapping genes, instead of
  the default behaviour, which is to maximize the total \underline{\emph{number}}
  of non-overlapping genes.

\exdent
  \verb`-n` \enskip or \enskip \verb`--no_header`

  Do not include the program-settings header information in the
  output file.  With this option, the output file will contain
  only the coordinates of the selected orfs.

\exdent
  \verb`-o` \Desc{n} \enskip or \enskip \verb`--max_olap` \Desc{n}

  Set the maximum overlap length to \Desc{n}.  Overlaps of this
  many or fewer bases between genes are not regarded as overlaps.

\exdent
  \verb`-t` \Desc{x} \enskip or \enskip \verb`--cutoff` \Desc{x}

  Only genes with an entropy distance score less than \Desc{x} will
  be considered.  This cutoff is made before any subsequent steps
  in the algorithm.

\exdent
  \verb`-w` \enskip or \enskip \verb`--without_stops`

  Do \underline{\emph{NOT}} include the stop codon in the region
  described by the output coordinates.  By default it is included.

\exdent
  \verb`-z` \Desc{n} \enskip or \enskip \verb`--trans_table` \Desc{n}

  Use Genbank translation table number \Desc{n} to specify stop codons.

\exdent
  \verb`-Z` \Desc{codon-list} \enskip or \enskip \verb`--stop_codons` \Desc{codon-list}

  Specify allowable stop codons as a comma-separated list.
  Sample format:  \verb`-Z tag,tga`.
  The default stop codons are \Pg{tag}, \Pg{tga} and \Pg{taa}.
\el

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Versions}

\subsection{Version~0.10}
  \bi\RaggedRight
  \item    
    Initial release.
  \ei

\raggedright
\bibliographystyle{alpha}
\bibliography{notes}

\end{document}
